includes/class-sensei-course-access.php

```diff
@@ -0,0 +1,294 @@
+
+	/**
+	 * Courses instances.
+	 *
+	 * @var static[]
+	 */
+	private static $instances = [];
+
+	/**
+	 * Course access providers.
+	 *
+	 * @var Sensei_Course_Access_Provider_Interface[]
+	 */
+	private static $access_providers;
+
+	/**
+	 * Course access providers version hash.
+	 *
+	 * @var string
+	 */
+	private static $access_providers_version;
+
+	/**
+	 * Course ID to handle access checks for.
+	 *
+	 * @var int
+	 */
+	private $course_id;
+
+	/**
+	 * Get singleton instance.
+	 *
+	 * @param int $course_id Course ID to handle checks for.
+	 *
+	 * @return self
+	 */
+	public static function get_course_instance( $course_id ) {
+		if ( ! isset( self::$instances[ $course_id ] ) ) {
+			self::$instances[ $course_id ] = new static( $course_id );
+		}
+
+		return self::$instances[ $course_id ];
+	}
+
+	/**
+	 * Sensei_Course_Access constructor.
+	 *
+	 * @param int $course_id Course ID to handle checks for.
+	 */
+	private function __construct( $course_id ) {
+		$this->course_id = $course_id;
+	}
+
+	/**
+	 * Return if a user has access.
+	 *
+	 * @param int  $user_id       User ID.
+	 * @param bool $force_recheck Force a recalculation of access.
+	 * @return bool
+	 */
+	public function has_access( $user_id, $force_recheck = false ) {
+		try {
+			if ( ! $force_recheck ) {
+				$access_check_log = $this->get_access_check_log( $user_id );
+				if (
+					$access_check_log
+					&& $access_check_log->get_version() === $this->get_course_access_providers_version()
+				) {
+					return $this->get_stored_access( $user_id );
+				}
+			}
+
+			$access_check_log = $this->check_access( $user_id );
+			$has_access       = $access_check_log->has_access();
+			if ( ! is_bool( $has_access ) ) {
+				$has_access = $this->get_default_access( $user_id );
+			}
+
+			$this->save_access( $user_id, $has_access );
+		} catch ( Exception $e ) {
+			$has_access = false;
+		}
+
+		return $has_access;
+	}
+
+	/**
+	 * Trigger course access check when access might have changed.
+	 *
+	 * @param int $user_id User ID.
+	 */
+	public function trigger_course_access_check( $user_id ) {
+		$this->has_access( $user_id, true );
+	}
+
+	/**
+	 * Get access from taxonomy record.
+	 *
+	 * @param int $user_id User ID.
+	 * @return bool
+	 * @throws Exception When learner term could not be created.
+	 */
+	private function get_stored_access( $user_id ) {
+		$term = Sensei_Learner::get_learner_term( $user_id );
+
+		return has_term( $term->term_id, 'learner', $this->course_id );
+	}
+
+	/**
+	 * Save access in taxonomy.
+	 *
+	 * @param int  $user_id    User ID.
+	 * @param bool $has_access If the user has access to the course.
+	 * @return bool
+	 * @throws Exception When learner term could not be created.
+	 */
+	private function save_access( $user_id, $has_access ) {
+		$term = Sensei_Learner::get_learner_term( $user_id );
+		if ( ! $has_access ) {
+			$result = wp_remove_object_terms( $this->course_id, [ intval( $term->term_id ) ], 'learner' );
+
```

Comment:
Convo with Alex: Possibly add filter to remove progress when they don't have access.